File: src/utils/clipboard.ts
import { execFileSync } from "child_process";

export type ClipboardCopyResult = { ok: true } | { ok: false; message: string };

interface ClipboardCommandOptions {
  input?: string;
  env?: NodeJS.ProcessEnv;
}

interface ClipboardDeps {
  platform: NodeJS.Platform;
  commandExists: (command: string) => boolean;
  runCommand: (command: string, args: string[], options?: ClipboardCommandOptions) => void;
  env: NodeJS.ProcessEnv;
}

function getErrorMessage(error: unknown): string {
  if (error instanceof Error) return error.message;
  return String(error);
}

function defaultRunCommand(command: string, args: string[], options?: ClipboardCommandOptions): void {
  execFileSync(command, args, {
    encoding: "utf-8",
    stdio: ["pipe", "pipe", "pipe"],
    input: options?.input,
    env: options?.env,
  });
}

function defaultCommandExists(command: string): boolean {
  try {
    if (process.platform === "win32") {
      execFileSync("where", [command], { stdio: "ignore" });
      return true;
    }

    execFileSync("which", [command], { stdio: "ignore" });
    return true;
  } catch {
    return false;
  }
}

export function copyTextToClipboardWithDeps(
  text: string,
  deps: ClipboardDeps,
): ClipboardCopyResult {
  try {
    if (deps.platform === "darwin") {
      if (!deps.commandExists("pbcopy")) {
        return { ok: false, message: "Clipboard backend 'pbcopy' is not available." };
      }

      deps.runCommand("pbcopy", [], { input: text, env: deps.env });
      return { ok: true };
    }

    if (deps.platform === "linux") {
      const candidates: Array<{ command: string; args: string[] }> = [
        { command: "wl-copy", args: [] },
        { command: "xclip", args: ["-selection", "clipboard"] },
        { command: "xsel", args: ["--clipboard", "--input"] },
      ];

      const available = candidates.filter((candidate) => deps.commandExists(candidate.command));
      if (available.length === 0) {
        return {
          ok: false,
          message: "No clipboard utility found (tried wl-copy, xclip, xsel).",
        };
      }

      let lastError = "Clipboard command failed.";
      for (const candidate of available) {
        try {
          deps.runCommand(candidate.command, candidate.args, { input: text, env: deps.env });
          return { ok: true };
        } catch (error) {
          lastError = getErrorMessage(error);
        }
      }

      return { ok: false, message: lastError };
    }

    if (deps.platform === "win32") {
      const shells = ["powershell.exe", "powershell", "pwsh.exe", "pwsh"];
      const shell = shells.find((candidate) => deps.commandExists(candidate));
      if (!shell) {
        return {
          ok: false,
          message: "No PowerShell executable found for clipboard access.",
        };
      }

      deps.runCommand(
        shell,
        ["-NoProfile", "-NonInteractive", "-Command", "Set-Clipboard -Value $env:TUI_MOODLE_CLIPBOARD_TEXT"],
        {
          env: {
            ...deps.env,
            TUI_MOODLE_CLIPBOARD_TEXT: text,
          },
        },
      );
      return { ok: true };
    }

    return { ok: false, message: `Clipboard copy is not supported on '${deps.platform}'.` };
  } catch (error) {
    return { ok: false, message: getErrorMessage(error) };
  }
}

export function copyTextToClipboard(text: string): ClipboardCopyResult {
  return copyTextToClipboardWithDeps(text, {
    platform: process.platform,
    commandExists: defaultCommandExists,
    runCommand: defaultRunCommand,
    env: process.env,
  });
}

File: src/platform/clipboard.rs
use crate::platform::{OpResult, command_exists};
use std::io::Write;
use std::process::{Command, Stdio};

pub fn copy_text(text: &str) -> OpResult {
    if cfg!(target_os = "macos") {
        if !command_exists("pbcopy") {
            return Err("Clipboard backend 'pbcopy' is not available.".to_owned());
        }
        return run_with_input("pbcopy", &[], text);
    }

    if cfg!(target_os = "linux") {
        let candidates: &[(&str, &[&str])] = &[
            ("wl-copy", &[]),
            ("xclip", &["-selection", "clipboard"]),
            ("xsel", &["--clipboard", "--input"]),
        ];
        let available: Vec<&(&str, &[&str])> =
            candidates.iter().filter(|(cmd, _)| command_exists(cmd)).collect();
        if available.is_empty() {
            return Err("No clipboard utility found (tried wl-copy, xclip, xsel).".to_owned());
        }
        let mut last_error = "Clipboard command failed.".to_owned();
        for (cmd, args) in available {
            match run_with_input(cmd, args, text) {
                Ok(()) => return Ok(()),
                Err(error) => last_error = error,
            }
        }
        return Err(last_error);
    }

    if cfg!(windows) {
        let shells = ["powershell.exe", "powershell", "pwsh.exe", "pwsh"];
        let shell = shells.iter().copied().find(|c| command_exists(c));
        let shell = match shell {
            Some(value) => value,
            None => return Err("No PowerShell executable found for clipboard access.".to_owned()),
        };
        let mut command = Command::new(shell);
        command
            .args([
                "-NoProfile",
                "-NonInteractive",
                "-Command",
                "Set-Clipboard -Value $env:TUI_MOODLE_CLIPBOARD_TEXT",
            ])
            .env("TUI_MOODLE_CLIPBOARD_TEXT", text);
        let output = command.output().map_err(|e| e.to_string())?;
        if !output.status.success() {
            return Err(String::from_utf8_lossy(&output.stderr).trim().to_owned());
        }
        return Ok(());
    }

    Err(format!("Clipboard copy is not supported on '{}'.", std::env::consts::OS))
}

fn run_with_input(command: &str, args: &[&str], input: &str) -> OpResult {
    let mut child = Command::new(command)
        .args(args)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .map_err(|e| e.to_string())?;
    if let Some(stdin) = child.stdin.as_mut() {
        stdin.write_all(input.as_bytes()).map_err(|e| e.to_string())?;
    }
    let output = child.wait_with_output().map_err(|e| e.to_string())?;
    if !output.status.success() {
        return Err(String::from_utf8_lossy(&output.stderr).trim().to_owned());
    }
    Ok(())
}
